Export Image Development (#80)
IBM/packer-plugin-ibmcloud: post-processor/README.md

## Description
The IBM Image Exporter post-processor will allow a customer to export to a COS bucket any VPC custom image they are authorized to that has become available and is not yet in the process of being deleted -- i.e. with status other than pending, failed, or deleting. 

Generated **Image** from **Packer Builder** is exported to **IBM COS Bucker** if **image_id** is not specified in the post processor block.
Also note if **image_id** is passed then **api_key** and **region** is mandatory.

### Post-Processor
- [ibmcloud-export](post-processor/ibmcloud-export) - The `ibmcloud-export` post-processor supports exporting custom images to COS bucket. 

### Prerequisites
Please refer to https://github.com/IBM/packer-plugin-ibmcloud/blob/5fe6132a8fd3d17e6e75c798a8ed7883496c25d5/README.md README.md file from the main repository section.

## Usage

Follow usage section of the README.md file till step 3 and follow the below step 4.

4. Finally, run Packer plugin commands
    ```shell
    packer validate examples/postprocessor.vpc.rhel.pkr.hcl
    packer build examples/postprocessor.vpc.rhel.pkr.hcl
    ```

***********

## Packer Template in detail

> Packer's behavior is determined by the Packer template. This template tells Packer not only the plugins (builders, provisioners, post-processors) to use, but also how to configure them and in what order run them.

Reviewer: Can we shuffle the order of this data structures of packer ?
like post processor>builder>provisioner ?

Author: No

Author: What ever order have specified in example folder

      image_id              = "r006-47ce771b-2de5-4f63-9606-c9b7421d6888"
      image_export_job_name = "image-export-packer-2"
      api_key               = "${var.ibm_api_key}"
      region                = "us-south"
      storage_bucket_name   = "storage-bucket-2"
      format                = "qcow2"
    }
    post-processor "ibmcloud-export" {
      image_id              = "r006-47ce771b-2de5-4f63-9606-c9b7421d6888"
      image_export_job_name = "image-export-packer-3"
      api_key               = "${var.ibm_api_key}"
      region                = "us-south"
      storage_bucket_name   = "storage-bucket-2"
      format                = "qcow2"
    }
  }
}
```
### Understanding Packer Template Blocks

#### `post-processor` Block
The `variable` block defines variables within your Packer configuration. Input variables serve as parameters for a Packer build, allowing aspects of the build to be customized without altering the build's own source code. When you declare variables in the build of your configuration, you can set their values using CLI options and environment variables.

Variable | Type |Description
--- | --- | ---
**post-processor export image variables** |
| |

> api_key | string | The IBM Cloud platform API key. Required.

Reviewer: This is optional, but required only if image-id is given right ?

Author: Done

> region | string | IBM Cloud region where VPC is deployed. Required.

Reviewer: same as above, not required.

Author: Done

      format                = "qcow2"
    }
  }
}

```

### Understanding Packer Template Blocks

#### `post-processor` Block
The `variable` block defines variables within your Packer configuration. Input variables serve as parameters for a Packer build, allowing aspects of the build to be customized without altering the build's own source code. When you declare variables in the build of your configuration, you can set their values using CLI options and environment variables.

Variable | Type |Description
--- | --- | ---
**post-processor export image variables** |
| |
api_key | string | The IBM Cloud platform API key. Required only if image_id is provided.
region | string | IBM Cloud region where VPC is deployed. Required only if image_id is provided.
vpc_endpoint_url | string | Configure URL for VPC test environments. Optional.
iam_url | string | Configure URL for IAM test environments. Optional.
format | string | The format to use for the exported image. If the image is encrypted, only qcow2 is supported. Only [ qcow2, vhd ] are supported. Optional.
image_id | string | The image identifier to export image. If unspecified builder image_id will be used. Optional. 

> image_export_job_name | string | The name for this image export job. Optional.

Author: Since I have taken out this logic as the api spec already has this feature implemented, this change is not required

# IBM Packer Plugin - Post-Processors

## Scope

> The IBM Packer Plugin Post Processor Plugin can be used to Export Images on IBM Cloud.

Reviewer: ```suggestion
The IBM Packer Plugin's export-image post processor can be used to export images on IBM Cloud.
```

## Description

> The IBM Image Exporter post-processor will allow a customer to export to a COS bucket any VPC custom image they are authorized to that has become available and is not yet in the process of being deleted -- i.e. with status other than pending, failed, or deleting. 

Reviewer: ```suggestion
IBM Cloud's export-image post-processor will allow a customer to export any VPC custom image to a COS bucket; When below conditions are followed: 

  -   User Authorization
  -   Image fall in same region and is in available state (with status other than pending, failed, or deleting.)

```

> Generated **Image** from **Packer Builder** is exported to **IBM COS Bucker** if **image_id** is not specified in the post processor block.

Reviewer: ```suggestion
Generated **Image** from **Packer Builder** will be exported to **IBM COS Bucket** optionally **image_id** can also be specified in the post processor block for exporting.
```

Also note if **image_id** is passed then **api_key** and **region** is mandatory.

### Post-Processor
- [ibmcloud-export-image](post-processor/ibmcloud-export-image) - The `ibmcloud-export-image` post-processor supports exporting custom images to COS bucket. 

### Prerequisites

> Please refer to https://github.com/IBM/packer-plugin-ibmcloud/blob/5fe6132a8fd3d17e6e75c798a8ed7883496c25d5/README.md README.md file from the main repository section.

Reviewer: ```suggestion
Please refer to [README.md](https://github.com/IBM/packer-plugin-ibmcloud/blob/master/README.md) file from the main repository section.
```

Correct link from main branch

## Usage

> Follow usage section of the README.md file from the main repository till step 3 and follow the below step 4.

Reviewer: ```suggestion
Follow [usage section](https://github.com/IBM/packer-plugin-ibmcloud/blob/master/README.md#:~:text=export%20OBJC_DISABLE_INITIALIZE_FORK_SAFETY%3DYES-,Usage,-Using%20the%20packer) of the README.md file from the main repository till step 3 and follow the below step 4.
```

Suggestion: Please reference the accurate link to the sub-section.

4. Finally, run Packer plugin commands
    ```shell
    packer validate examples/postprocessor.vpc.rhel.pkr.hcl
    packer build examples/postprocessor.vpc.rhel.pkr.hcl
    ```

***********

## Packer Template in detail
Follow Packer Template in detail section of the README.md file from the main repository and refer here for the post processor template.

> ```hcl

Reviewer: Please just keep the post processor section rather than repeating the builder component.
